alchemy/src/cloudflare/bucket.ts

```diff
@@ -472,144 +454,31 @@ export async function R2Bucket(
-          return result as unknown as R2ObjectContent;
-        }
-        return null;
-      }
-      const response = await getObject(api, {
-        bucketName: bucket.name,
-        key,
-      });
-      if (response.ok) {
-        return parseR2Object(key, response);
-      } else if (response.status === 404) {
-        return null;
-      } else {
-        throw await handleApiError(response, "get", "object", key);
-      }
-    },
-    list: async (options?: R2ListOptions): Promise<R2Objects> => {
-      if (isLocal) {
-        return (await localBucket()).list(options);
-      }
-      return listObjects(api, bucket.name, {
-        ...options,
-        jurisdiction: bucket.jurisdiction,
-      });
-    },
-    put: async (
-      key: string,
-      value: PutObjectObject,
-      options?: Pick<R2PutOptions, "httpMetadata">,
-    ): Promise<PutR2ObjectResponse> => {
-      if (isLocal) {
-        return await (await localBucket()).put(
-          key,
-          typeof value === "string"
-            ? value
-            : Buffer.isBuffer(value) ||
-                value instanceof Uint8Array ||
-                value instanceof ArrayBuffer
-              ? new Uint8Array(value)
-              : value instanceof Blob
-                ? new Uint8Array(await value.arrayBuffer())
-                : value instanceof ReadableStream
-                  ? new Uint8Array(await streamToBuffer(value))
-                  : value,
-          options,
-        );
-      }
-      const response = await putObject(api, {
-        bucketName: bucket.name,
-        key: key,
-        object: value,
-        options: options,
-      });
-      const body = (await response.json()) as {
-        result: {
-          key: string;
-          etag: string;
-          uploaded: string;
-          version: string;
-          size: string;
-        };
-      };
-      return {
-        key: body.result.key,
-        etag: body.result.etag,
-        uploaded: new Date(body.result.uploaded),
-        version: body.result.version,
-        size: Number(body.result.size),
-      };
-    },
-    delete: async (key: string) => {
-      if (isLocal) {
-        await (await localBucket()).delete(key);
-      }
-      return deleteObject(api, {
-        bucketName: bucket.name,
-        key: key,
-      });
+  return makeAsyncProxyForBinding({
+    apiOptions: props,
+    name: id,
+    binding: bucket as Omit<R2Bucket, keyof globalThis.R2Bucket>,
+    properties: {
+      createMultipartUpload: true,
+      delete: true,
+      get: true,
+      head: true,
+      list: true,
+      put: true,
+      resumeMultipartUpload: (promise) => (key: string, uploadId: string) =>
+        makeAsyncProxy(
+          { key, uploadId },
+          promise.then((bucket) => bucket.resumeMultipartUpload(key, uploadId)),
+          {
+            uploadPart: true,
+            abort: true,
+            complete: true,
+          },
+        ),
```

Comment:
`resumeMultipartUpload` does not return a promise. To match the type of the runtime API, we make this synchronous by returning another proxy.

This is also done for D1 prepared statements and sessions.